### DISCUSSION TITLE
failure to save metrics caused postgres pod to crash

### DISCUSSION THREAD

**Discussion by ns-jhu** [cloudnative-pg/cloudnative-pg#1651] [General]
I noticed this pod failure:

pg-introspect01-3                    0/1     Init:CrashLoopBackOff   18         14h
4

the associated error log is: 

{"level":"error","ts":1677863495.3143833,"msg":"Error while committing metrics extraction","logging_pod":"pg-introspect01-3","error":"commit unexpectedly resulted in rollback","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/log/log.go:127\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/log.Error\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/log/log.go:165\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.QueryCollector.collect.func1\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:329\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.QueryCollector.collect\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:335\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.(*QueriesCollector).collectUserQueries\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:129\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.QueriesCollector.Collect\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:64\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/webserver/metricserver.(*Exporter).collectPgMetrics\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/webserver/metricserver/pg_collector.go:340\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/webserver/metricserver.(*Exporter).Collect\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/webserver/metricserver/pg_collector.go:273\ngithub.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1\n\t/home/nsadmin/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/registry.go:456"}
{"level":"error","ts":1677863495.314449,"msg":"Error collecting user query","query":"pg_replication_slots","logging_pod":"pg-introspect01-3","targetDatabase":"app","error":"ERROR: recovery is in progress (SQLSTATE 55000)","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/log/log.go:127\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.(*QueriesCollector).collectUserQueries\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:131\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics.QueriesCollector.Collect\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/metrics/collector.go:64\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/webserver/metricserver.(*Exporter).collectPgMetrics\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/webserver/metricserver/pg_collector.go:340\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/webserver/metricserver.(*Exporter).Collect\n\t/Users/jinsong/Documents/netskope/cloudnative-pg.netskope/pkg/management/postgres/webserver/metricserver/pg_collector.go:273\ngithub.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1\n\t/home/nsadmin/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/registry.go:456"}

it seems the issue is that the scraper asked postgres to get some metrics. and when the metrics is saved to prometheus database and fail, the postgres pod crash. 

in reality, we can't expect the outside  prometheus system to always behave perfectly. even though metrics are important, they are not so important to justify postgres going down, which can bring much bigger impact. the code should be changed so that when metrics saving fails, it log this error, but database should continue to work.

**Comment by ns-jhu** [comment]
I made the code change in my private branch to keep pods running even with metrics failure, but after this, the pods still crashed under heavy write pressure, the error is:

C02C131VMD6T:cnpg-cluster jinsong$ kubectl logs -n tpdata pg-introspect01-3  -c bootstrap-controller
{"level":"info","ts":1678125748.822473,"msg":"Installing the manager executable","destination":"/controller/manager","version":"1.18.1","build":{"Version":"","Commit":"438991ae","Date":"2023-03-03"}}
panic: open /controller/manager: text file busy

goroutine 1 [running]:
github.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/bootstrap.NewCmd.func1(0xc00047c600?, {0xc0003e3da0?, 0x2?, 0x2?})
	/home/jhu/cloudnative-pg/internal/cmd/manager/bootstrap/cmd.go:44 +0x247
github.com/spf13/cobra.(*Command).execute(0xc00047c600, {0xc0003e3d80, 0x2, 0x2})
	/home/jhu/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc00047c000)
	/home/jhu/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/jhu/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
main.main()
	/home/jhu/cloudnative-pg/cmd/manager/main.go:64 +0x259
C02C131VMD6T:cnpg-cluster jinsong$ kubectl logs -n tpdata pg-introspect01-1  -c bootstrap-controller
{"level":"info","ts":1678092453.9648392,"msg":"Installing the manager executable","destination":"/controller/manager","version":"1.18.1","build":{"Version":"","Commit":"438991ae","Date":"2023-03-03"}}
{"level":"info","ts":1678092454.1175983,"msg":"Setting 0750 permissions"}
{"level":"info","ts":1678092454.1176631,"msg":"Bootstrap completed"}


I checked the process in the pod, it is:

postgres@pg-introspect01-1:/$ ps -ef | grep manager
postgres       1       0  2 08:47 ?        00:11:51 /controller/manager instance run --log-level=info
postgres  487450      38  0 18:11 ?        00:00:00 sh -c /controller/manager wal-restore --log-destination /controller/log/postgres.json 000000090000044100000013 pg_wal/RECOVERYXLOG
postgres  487451  487450  2 18:11 ?        00:00:00 /controller/manager wal-restore --log-destination /controller/log/postgres.json 000000090000044100000013 pg_wal/RECOVERYXLOG

**Comment by gbartolini** [comment]
Could you please share your cluster's YAML configuration file?

Thanks,
Gabriele

**Comment by mnencia** [comment]
The [latest development version](https://cloudnative-pg.io/documentation/1.19/installation_upgrade/#testing-the-latest-development-snapshot) contains a [fix](https://github.com/cloudnative-pg/cloudnative-pg/commit/29bfdcc8f43472c5264db401764f0f9300cf9134) that may solve your issue. Can you please test it?

**Reply by ns-jhu** [reply]
sure.  I will do it later. I also had implemented a more brute force fix  in pkg/management/postgres/metrics/collector.go, to ensure  function 

func (c QueryCollector) collect(conn *sql.DB, ch chan<- prometheus.Metric) error 

only logs error inside and always return nil . and  tested this solution, and found it solved the problem. 

looking at the purpose of the function, I don't see the why k8s prometheus glitch should cause postgres failure.  the bottom line is, whatever bad things happens to prometheus, the postgres should continue to run.  so probably we should combine your fix with mine. Here is my changed version:

// collect retrieves metrics from query and exposes them to prometheus
func (c QueryCollector) collect(conn *sql.DB, ch chan<- prometheus.Metric) error {
	tx, err := createMonitoringTx(conn)
	if err != nil {
		
		log.Error(err, "Error while createMonitoringTx ")
		return nil
	}

	defer func() {
		if err := tx.Commit(); err != nil {
			log.Error(err, "Error while committing metrics extraction")
		}
	}()

	rows, err := tx.Query(c.userQuery.Query)
	if err != nil {
		
		log.Error(err, "Error while tx.Query(c.userQuery.Query) ")
		return nil
	}
	defer func() {
		if err := rows.Close(); err != nil {
			log.Warning("Error while closing metrics extraction",
				"err", err.Error())
		}
		if err := rows.Err(); err != nil {
			log.Warning("Error while loading metrics",
				"err", err.Error())
		}
	}()

	columns, err := rows.Columns()
	if err != nil {
		
		log.Error(err, "Error while rows.Columns() ")
		return nil
	}

	columnData := make([]interface{}, len(columns))
	scanArgs := make([]interface{}, len(columns))
	for i := range columnData {
		scanArgs[i] = &columnData[i]
	}

	if len(columns) != len(c.columnMapping) {
		log.Warning("Columns number mismatch",
			"name", c.namespace,
			"columnNumberFromDB", len(columns),
			"columnNumberFromConfiguration", len(c.columnMapping))
		return nil
	}

	for rows.Next() {
		if err = rows.Scan(scanArgs...); err != nil {
			
			log.Error(err, "Error while rows.Scan(scanArgs...)")
			return nil
		}

		labels, done := c.collectLabels(columns, columnData)
		if done {
			c.collectColumns(columns, columnData, labels, ch)
		}
	}
	return nil
}

**Reply by ns-jhu** [reply]
for the issue 
panic: open /controller/manager: text file busy

I made the change in internal/cmd/manager/bootstrap/cmd.go to check if the file exist or not first, and only copy if the file is not there first, and that solved the problem. Here is the code after I changed:

func NewCmd() *cobra.Command {
	cmd := cobra.Command{
		Use:  "bootstrap [target]",
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			dest := args[0]

			log.Info("Installing the manager executable",
				"destination", dest,
				"version", versions.Version,
				"build", versions.Info)
			exists, err := fileutils.FileExists(dest)
			if err != nil {
				panic(err)
			}
			if !exists {
				err = fileutils.CopyFile(cmd.Root().Name(), dest)
				if err != nil {
					panic(err)
				}
			} else {
				log.Info(dest + "already exist")
			}

			log.Info("Setting 0750 permissions")
			err = os.Chmod(dest, 0o750) // #nosec
			if err != nil {
				panic(err)
			}

			log.Info("Bootstrap completed")

			return nil
		},
	}

	return &cmd
}


I have tested both of my changes  under stress testing for 3 days, and confirmed that they work fine.

**Comment by ns-jhu** [comment]
@mnencia, can you help to add my change to the code base ? I tried to create a PR, but found that I don't have permission. 
Here is my changed code for internal/cmd/manager/bootstrap/cmd.go

/*
Copyright The CloudNativePG Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package bootstrap implement the "controller bootstrap" command
package bootstrap

import (
        "os"

        "github.com/spf13/cobra"

        "github.com/cloudnative-pg/cloudnative-pg/pkg/fileutils"
        "github.com/cloudnative-pg/cloudnative-pg/pkg/management/log"
        "github.com/cloudnative-pg/cloudnative-pg/pkg/versions"
)

// NewCmd create a new cobra command
func NewCmd() *cobra.Command {
        cmd := cobra.Command{
                Use:  "bootstrap [target]",
                Args: cobra.ExactArgs(1),
                RunE: func(cmd *cobra.Command, args []string) error {
                        dest := args[0]

                        log.Info("Installing the manager executable",
                                "destination", dest,
                                "version", versions.Version,
                                "build", versions.Info)
                        exists, err := fileutils.FileExists(dest)
                        if err != nil {
                                panic(err)
                        }
                        if !exists {
                                err = fileutils.CopyFile(cmd.Root().Name(), dest)
                                if err != nil {
                                        panic(err)
                                }
                        } else {
                                log.Info(dest + "already exist")
                        }

                        log.Info("Setting 0750 permissions")
                        err = os.Chmod(dest, 0o750) // #nosec
                        if err != nil {
                                panic(err)
                        }

                        log.Info("Bootstrap completed")

                        return nil
                },
        }

        return &cmd
}






here is my changed code for pkg/management/postgres/metrics/collector.go



/*
Copyright The CloudNativePG Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// This code is inspired on [postgres_exporter](https://github.com/prometheus-community/postgres_exporter)

package metrics

import (
        "context"
        "database/sql"
        "fmt"
        "path"
        "regexp"

        "github.com/blang/semver"
        "github.com/prometheus/client_golang/prometheus"

        "github.com/cloudnative-pg/cloudnative-pg/pkg/management/log"
        "github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres"
        "github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/metrics/histogram"
        postgresutils "github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres/utils"
)

// QueriesCollector is the implementation of PgCollector for a certain
// collection of custom queries supplied by the user
type QueriesCollector struct {
        instance      *postgres.Instance
        defaultDBName string
        collectorName string

        userQueries    UserQueries
        mappings       map[string]MetricMapSet
        variableLabels map[string]VariableSet

        errorUserQueries      *prometheus.CounterVec
        errorUserQueriesGauge prometheus.Gauge
}

// Name returns the name of this collector, as supplied by the user in the configMap
func (q QueriesCollector) Name() string {
        return q.collectorName
}

var isPathPattern = regexp.MustCompile(`[][*?]`)

// Collect loads data from the actual PostgreSQL instance
func (q QueriesCollector) Collect(ch chan<- prometheus.Metric) error {
        // Reset before collecting
        q.errorUserQueries.Reset()

        err := q.collectUserQueries(ch)
        if err != nil {
                return err
        }

        // Add errors into errorUserQueriesVec and errorUserQueriesGauge metrics
        q.errorUserQueriesGauge.Collect(ch)
        q.errorUserQueries.Collect(ch)

        return nil
}

func (q *QueriesCollector) collectUserQueries(ch chan<- prometheus.Metric) error {
        isPrimary, err := q.instance.IsPrimary()
        if err != nil {
                return err
        }

        // In case more than one user query specify a pattern in target_databases,
        // we need to get them just once
        var allAccessibleDatabasesCache []string

        for name, userQuery := range q.userQueries {
                queryLogger := log.WithValues("query", name)
                collector := QueryCollector{
                        namespace:      name,
                        userQuery:      userQuery,
                        columnMapping:  q.mappings[name],
                        variableLabels: q.variableLabels[name],
                }

                if !q.toBeChecked(name, userQuery, isPrimary, queryLogger) {
                        continue
                }

                queryLogger.Debug("Collecting data")

                targetDatabases := userQuery.TargetDatabases
                if len(targetDatabases) == 0 {
                        targetDatabases = append(targetDatabases, q.defaultDBName)
                }

                // Initialize the cache is one of the target contains a pattern
                if allAccessibleDatabasesCache == nil {
                        for _, targetDatabase := range targetDatabases {
                                if isPathPattern.MatchString(targetDatabase) {
                                        databases, err := q.getAllAccessibleDatabases()
                                        if err != nil {
                                                q.reportUserQueryErrorMetric(name + ": " + err.Error())
                                                break
                                        }
                                        allAccessibleDatabasesCache = databases
                                        break
                                }
                        }
                }

                allTargetDatabases := q.expandTargetDatabases(targetDatabases, allAccessibleDatabasesCache)
                for targetDatabase := range allTargetDatabases {
                        conn, err := q.instance.ConnectionPool().Connection(targetDatabase)
                        if err != nil {
                                q.reportUserQueryErrorMetric(name + ": " + err.Error())
                                continue
                        }

                        err = collector.collect(conn, ch)
                        if err != nil {
                                queryLogger.Error(err, "Error collecting user query",
                                        "targetDatabase", targetDatabase)
                                // Increment metrics counters.
                                q.reportUserQueryErrorMetric(name + " on db " + targetDatabase + ": " + err.Error())
                        }
                }
        }
        return nil
}

func (q QueriesCollector) toBeChecked(name string, userQuery UserQuery, isPrimary bool, queryLogger log.Logger) bool {
        if (userQuery.Primary || userQuery.Master) && !isPrimary { // wokeignore:rule=master
                queryLogger.Debug("Skipping because runs only on primary")
                return false
        }

        if runOnServer := userQuery.RunOnServer; runOnServer != "" {
                matchesVersion, err := q.checkRunOnServerMatches(runOnServer, name)
                // any error should result in the query not being executed
                if err != nil {
                        queryLogger.Error(err, "while checking runOnServer version matches",
                                "runOnServer", runOnServer)
                        q.reportUserQueryErrorMetric(name)
                        return false
                } else if !matchesVersion {
                        queryLogger.Debug("Skipping because runs only on other postgres versions",
                                "runOnServer", runOnServer)
                        return false
                }
        }

        return true
}

func (q QueriesCollector) reportUserQueryErrorMetric(label string) {
        q.errorUserQueries.WithLabelValues(label).Inc()
        q.errorUserQueriesGauge.Set(1)
}

func (q QueriesCollector) checkRunOnServerMatches(runOnServer string, name string) (bool, error) {
        // The instance will only get the PostgreSQL version one time
        // and then cache the result, so this is not a real query
        pgVersion, err := q.instance.GetPgVersion()
        if err != nil {
                log.Error(err, "while parsing runOnServer queries")
                return false, err
        }

        isVersionInRange, err := semver.ParseRange(runOnServer)
        if err != nil {
                log.Error(err, "while parsing runOnServer version range",
                        "runOnServer", runOnServer, "query", name)
                return false, err
        }

        return isVersionInRange(pgVersion), nil
}

func (q QueriesCollector) expandTargetDatabases(
        targetDatabases []string,
        allAccessibleDatabasesCache []string,
) (allTargetDatabases map[string]bool) {
        allTargetDatabases = make(map[string]bool)
        for _, targetDatabase := range targetDatabases {
                if !isPathPattern.MatchString(targetDatabase) {
                        allTargetDatabases[targetDatabase] = true
                        continue
                }
                for _, database := range allAccessibleDatabasesCache {
                        matched, err := path.Match(targetDatabase, database)
                        if err == nil && matched {
                                allTargetDatabases[database] = true
                        }
                }
        }
        return allTargetDatabases
}

func (q QueriesCollector) getAllAccessibleDatabases() ([]string, error) {
        conn, err := q.instance.ConnectionPool().Connection(q.defaultDBName)
        if err != nil {
                return nil, fmt.Errorf("while connecting to expand target_database *: %w", err)
        }
        tx, err := createMonitoringTx(conn)
        if err != nil {
                return nil, fmt.Errorf("while creating monitoring tx to retrieve accessible databases list: %w", err)
        }
        defer func() {
                if err := tx.Commit(); err != nil {
                        log.Error(err, "Error while committing monitoring tx to retrieve accessible databases list")
                }
        }()
        databases, errors := postgresutils.GetAllAccessibleDatabases(tx, "datallowconn AND NOT datistemplate")
        if errors != nil {
                return nil, fmt.Errorf("while discovering databases for metrics: %v", errors)
        }
        return databases, nil
}

// Describe implements the prometheus.Collector and defines the metrics with return
func (q QueriesCollector) Describe(ch chan<- *prometheus.Desc) {
        for name, userQuery := range q.userQueries {
                collector := QueryCollector{
                        namespace:     name,
                        userQuery:     userQuery,
                        columnMapping: q.mappings[name],
                }

                collector.describe(ch)
        }

        // add error user queries description
        q.errorUserQueries.Describe(ch)
        q.errorUserQueriesGauge.Describe(ch)
}

// NewQueriesCollector creates a new PgCollector working over a set of custom queries
// supplied by the user
func NewQueriesCollector(
        name string,
        instance *postgres.Instance,
        defaultDBName string,
) *QueriesCollector {
        return &QueriesCollector{
                collectorName:  name,
                instance:       instance,
                mappings:       make(map[string]MetricMapSet),
                variableLabels: make(map[string]VariableSet),
                userQueries:    make(UserQueries),
                defaultDBName:  defaultDBName,
                errorUserQueries: prometheus.NewCounterVec(prometheus.CounterOpts{
                        Namespace: name,
                        Name:      "errors_total",
                        Help:      "Total errors occurred performing user queries.",
                }, []string{"errorUserQueries"}),
                errorUserQueriesGauge: prometheus.NewGauge(prometheus.GaugeOpts{
                        Namespace: name,
                        Name:      "last_error",
                        Help:      "1 if the last collection ended with error, 0 otherwise.",
                }),
        }
}

// ParseQueries parses a YAML file containing custom queries and add it
// to the set of gathered one
func (q *QueriesCollector) ParseQueries(customQueries []byte) error {
        var err error

        parsedQueries, err := ParseQueries(customQueries)
        if err != nil {
                return err
        }
        for name, query := range parsedQueries {
                if _, found := q.userQueries[name]; found {
                        log.Warning("Query with the same name already found. Overwriting the existing one.",
                                "queryName",
                                name)
                }

                q.userQueries[name] = query
                q.mappings[name], q.variableLabels[name] = query.ToMetricMap(
                        fmt.Sprintf("%v_%v", q.collectorName, name))
        }

        return nil
}

// InjectUserQueries injects the passed queries
func (q *QueriesCollector) InjectUserQueries(defaultQueries UserQueries) {
        if q == nil {
                return
        }

        for name, query := range defaultQueries {
                q.userQueries[name] = query
                q.mappings[name], q.variableLabels[name] = query.ToMetricMap(
                        fmt.Sprintf("%v_%v", q.collectorName, name))
        }
}

// QueryCollector is the implementation of PgCollector for a certain
// custom query supplied by the user
type QueryCollector struct {
        namespace      string
        userQuery      UserQuery
        columnMapping  MetricMapSet
        variableLabels VariableSet
}

// collect retrieves metrics from query and exposes them to prometheus
func (c QueryCollector) collect(conn *sql.DB, ch chan<- prometheus.Metric) error {
        tx, err := createMonitoringTx(conn)
        if err != nil {
                log.Error(err, "Error while createMonitoringTx ")
                return nil
        }

        defer func() {
                if err := tx.Commit(); err != nil {
                        log.Error(err, "Error while committing metrics extraction")
                }
        }()

        rows, err := tx.Query(c.userQuery.Query)
        if err != nil {
                log.Error(err, "Error while tx.Query(c.userQuery.Query) ")
                return nil
        }
        defer func() {
                if err := rows.Close(); err != nil {
                        log.Warning("Error while closing metrics extraction",
                                "err", err.Error())
                }
                if err := rows.Err(); err != nil {
                        log.Warning("Error while loading metrics",
                                "err", err.Error())
                }
        }()

        columns, err := rows.Columns()
        if err != nil {
                log.Error(err, "Error while rows.Columns() ")
                return nil
        }

        columnData := make([]interface{}, len(columns))
        scanArgs := make([]interface{}, len(columns))
        for i := range columnData {
                scanArgs[i] = &columnData[i]
        }

        if len(columns) != len(c.columnMapping) {
                log.Warning("Columns number mismatch",
                        "name", c.namespace,
                        "columnNumberFromDB", len(columns),
                        "columnNumberFromConfiguration", len(c.columnMapping))
                return nil
        }

        for rows.Next() {
                if err = rows.Scan(scanArgs...); err != nil {
                        log.Error(err, "Error while rows.Scan(scanArgs...)")
                        return nil
                }

                labels, done := c.collectLabels(columns, columnData)
                if done {
                        c.collectColumns(columns, columnData, labels, ch)
                }
        }
        return nil
}

// Collect the list of labels from the database, and returns true if the
// label extraction succeeded, false otherwise
func (c QueryCollector) collectLabels(columns []string, columnData []interface{}) ([]string, bool) {
        var labels []string
        for idx, columnName := range columns {
                if mapping, ok := c.columnMapping[columnName]; ok && mapping.Label {
                        value, ok := postgresutils.DBToString(columnData[idx])
                        if !ok {
                                log.Warning("Label value cannot be converted to string",
                                        "value", value,
                                        "mapping", mapping)
                                return nil, false
                        }

                        labels = append(labels, value)
                }
        }
        return labels, true
}

// Collect the metrics from the database columns
func (c QueryCollector) collectColumns(columns []string, columnData []interface{},
        labels []string, ch chan<- prometheus.Metric,
) {
        for idx, columnName := range columns {
                mapping, ok := c.columnMapping[columnName]
                if !ok {
                        log.Warning("Missing mapping for column", "column", columnName, "mapping", c.columnMapping)
                        continue
                }

                // There is a strong difference between histogram and non-histogram metrics in
                // postgres_exporter. The first ones are looked up by column name and the second
                // ones are looked up just using the index.
                //
                // We implemented the same behavior here.

                switch {
                case mapping.Discard || mapping.Label:
                        continue

                case mapping.Histogram:
                        histogramData, err := histogram.NewFromRawData(columnData, columns, columnName)
                        if err != nil {
                                log.Error(err, "Cannot process histogram metric",
                                        "columns", columns,
                                        "columnName", columnName,
                                        "mapping.Name", mapping.Name,
                                        "mappings", c.columnMapping,
                                        "mapping", mapping,
                                        "columnData", columnData,
                                        "labels", labels)
                        } else {
                                c.collectHistogramMetric(mapping, histogramData, labels, ch)
                        }

                default:
                        c.collectConstMetric(mapping, columnData[idx], labels, ch)
                }
        }
}

// createMonitoringTx create a monitoring transaction with read-only access
// and role set to `pg_monitor`
func createMonitoringTx(conn *sql.DB) (*sql.Tx, error) {
        tx, err := conn.BeginTx(context.Background(), &sql.TxOptions{
                ReadOnly: true,
        })
        if err != nil {
                return nil, err
        }

        defer func() {
                if err != nil {
                        _ = tx.Rollback()
                }
        }()

        _, err = tx.Exec("SET application_name TO cnpg_metrics_exporter")
        if err != nil {
                return nil, err
        }

        _, err = tx.Exec("SET ROLE TO pg_monitor")

        return tx, err
}

// describe puts in the channel the metadata we have for the queries we collect
func (c QueryCollector) describe(ch chan<- *prometheus.Desc) {
        for _, mapSet := range c.columnMapping {
                ch <- mapSet.Desc
        }
}

// collectConstMetric reports to the prometheus library a constant metric
func (c QueryCollector) collectConstMetric(
        mapping MetricMap, value interface{}, variableLabels []string, ch chan<- prometheus.Metric,
) {
        if mapping.Conversion == nil {
                log.Warning("Missing conversion while parsing value",
                        "namespace", c.namespace,
                        "value", value,
                        "mapping", mapping)
                return
        }

        floatData, ok := mapping.Conversion(value)
        if !ok {
                log.Warning("Error while parsing value",
                        "namespace", c.namespace,
                        "value", value,
                        "mapping", mapping)
                return
        }

        // Generate the metric
        metric, err := prometheus.NewConstMetric(mapping.Desc, mapping.Vtype, floatData, variableLabels...)
        if err != nil {
                log.Error(err, "while collecting constant metric", "metric", mapping.Name)
                return
        }
        ch <- metric
}

// collectHistogramMetric reports to the prometheus library an histogram-based metric
func (c QueryCollector) collectHistogramMetric(
        mapping MetricMap,
        columnData *histogram.Value,
        variableLabels []string,
        ch chan<- prometheus.Metric,
) {
        metric, err := prometheus.NewConstHistogram(
                mapping.Desc,
                columnData.Count, columnData.Sum, columnData.Buckets,
                variableLabels...,
        )
        if err != nil {
                log.Error(err, "while collecting histogram metric", "metric", mapping.Name)
                return
        }
        ch <- metric
}

**Comment by ns-jhu** [comment]
I created the PR https://github.com/cloudnative-pg/cloudnative-pg/pull/1905/files , can somebody review ?
